From: Jiri Pirko <jpirko@redhat.com>
Date: Wed, 17 Feb 2010 09:28:39 +0100
Subject: [fs] ext4: Avoid null pointer dereference when decoding EROFS w/o a journal
Message-id: 20100217082838.GB3360@psychotron.redhat.com
O-Subject: [RHEL5.4.z patch] CVE-2009-4308 BZ547256 ext4: Avoid null pointer dereference when decoding EROFS w/o a journal
Bugzilla: 547256
CVE: CVE-2009-4308
RH-Acked-by: Jerome Marchand <jmarchan@redhat.com>
RH-Acked-by: Anton Arapov <anton@redhat.com>
RH-Acked-by: Larry Woodman <lwoodman@redhat.com>
RH-Acked-by: Eric Sandeen <sandeen@redhat.com>

CVE-2009-4308
BZ547256
https://bugzilla.redhat.com/show_bug.cgi?id=547256

Description:
We need to check to make sure a journal is present before checking the
journal flags in ext4_decode_error().

This fix was included in rhel5.5 as a part of big ext4 rebase.

Upstream:
http://git.kernel.org/?p=linux/kernel/git/torvalds/linux-2.6.git;a=commitdiff;h=78f1ddbb498283c2445c11b0dfa666424c301803

Brew:
https://brewweb.devel.redhat.com/taskinfo?taskID=2263576

Jirka

Signed-off-by: Jiri Pirko <jpirko@redhat.com>

diff --git a/fs/ext4/super.c b/fs/ext4/super.c
index 7e02500..d3be1e2 100644
--- a/fs/ext4/super.c
+++ b/fs/ext4/super.c
@@ -336,7 +336,8 @@ static const char *ext4_decode_error(struct super_block *sb, int errno,
 		errstr = "Out of memory";
 		break;
 	case -EROFS:
-		if (!sb || EXT4_SB(sb)->s_journal->j_flags & JBD2_ABORT)
+		if (!sb || (EXT4_SB(sb)->s_journal &&
+			    EXT4_SB(sb)->s_journal->j_flags & JBD2_ABORT))
 			errstr = "Journal has aborted";
 		else
 			errstr = "Readonly filesystem";

